### PR TITLE
DEP: Deprecate error clearing for special method in array-coercion

### LINKED ISSUE #18199: DEP: Deprecate error clearing in array-protocol attribute fetching
See gh-18197, we currently clear almost all errors that occur when fetching `obj.__array_interface__` or just `obj.__array__`. There is little use for this, since by definition `AttributeError` can be raised if an object actually wants the error to be ignored.  Shapely dependend on this (raising a `NotImplementedError`).

We should deprecate ignoring these errors entirely.  There are more similar cases, although this issue is about the interface/array path.

(Will assign myself, but if anyone wants to look at it, feel free)

### LINKED ISSUE #18723: np.array silently catching errors
I find that `np.array` silently catches an error raised by `__array_interface__`. This was confirmed to be a bug on Gitter - https://gitter.im/numpy/numpy?at=6067b4ef3153ce63a3b6dcad

### Reproducing code example:

```python
import numpy as np

class Sample:
    @property
    def __array_interface__(self):
        raise OSError

x = Sample()
try:
    x.__array_interface__
except OSError:
    print("OSError raised by __array_interface__")

np.array(x)
print("But not by np.array")
```

### Error message:

No error. That's the problem.

The output of my code example is
```
OSError raised by __array_interface__
But not by np.array
```

### NumPy/Python version information:

1.20.2 3.8.9 (default, Apr  3 2021, 19:49:13) 
[Clang 12.0.0 (clang-1200.0.32.29)]

### PR DESCRIPTION
**Pull request by seberg**
When looking up `__array__` and `__array_interface__`, previously
most errors were ignored (during the attribute lookup for __array__).

This deprecates this, in the future only AttributeError (which is
ignored in `LookupSpecial` itself) will be ignored.

Closes gh-18723

---

Closes gh-18199 ~I feel like there is probably at least one more issue open, but didn't find it.~

### CODE REVIEW

File: doc/release/upcoming_changes/19001.deprecation.rst [hunk 1]
```diff
@@ -0,0 +1,8 @@
+Errors will be raised during array-like creation
+------------------------------------------------
+When an object raised an error during access of the special
+attributes ``__array__`` or ``__array_interface__``, this error
+was usually ignored.
+A warning is now given when the error is anything but AttributeError.
+To silence the warning, the type raising a warning has to be adapted
```
Comment (eric-wieser) [inline doc/release/upcoming_changes/19001.deprecation.rst:7]:
```suggestion
To silence the warning, the type raising the exception has to be adapted
```

**Review by eric-wieser — APPROVED** [review]
Looks sensible, although it might be better to use the word "exception" rather than "error in a few places.

**Comment by seberg** [comment]
Well, that was confusing :).  I changed the error message (error -> exception), so the warning filter was not set up correctly in the test and thus the strange failure.

Should probably just squash merge this...

**Comment by charris** [comment]
Thanks Sebastian.

**Comment by radarhere** [comment]
Thanks. Just out of interest, is there a standard numpy timeline on moving from deprecation to exception?

**Comment by seberg** [comment]
@radarhere typically 2 releases, although its not uncommen we forgot, but after two releases we will definitely merge a PR unless there is some protest.  If there are *very* good reasons, I guess we could speed things up after some discussion.  (In this case, the only known issue was because `shapely` ran into this change, since shapely is fixed, I might consider it.  But it should still have a good reason of why it helps.)

**Pull request merged**